### ISSUE TITLE
Significant time difference of calculating Jacobian matrix using jacrev and oracle functions

### ISSUE THREAD

**Issue by FrankTianTT** [pytorch/pytorch#105220]
### 🚀 The feature, motivation and pitch

Sorry I'm not sure if this is a new feature, but I'd be very grateful if you guys would like to improve this part.

Specifically, I found that using `jacrev` would be much slower than using the oracle Jacobian function:
```python
from torch.func import vmap, jacrev
import torch
import time

a = torch.rand(10000, 10000)

def f(x):
    return (x ** 2).sum(-1)

def df(x):
    return 2 * x


t0 = time.time()
b = df(a)
t1 = time.time()
c = vmap(jacrev(f))(a)
t2= time.time()

assert torch.allclose(b, c)
print(t1 - t0, t2 - t1)
```
result: 0.10568618774414062 0.9206998348236084

Given that oracle's Jacobian is readily available in neural networks, I wonder why using `jacrev` is so much slower? Is there something wrong with me?

Of course, I can actually rewrite each layer of the neural network to obtain the value and Jacobian at the same time, but calculating the Hessian matrix is too troublesome. It would be great if `jacrev` could be faster!

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @Chillee @samdow @kshitij12345 @janeyx99

**Comment by zou3519** [comment]
The issue is that fundamentally, `vmap(jacrev(f))(a)` needs to run the forward pass first to generate an autograd graph before it can run the backward pass to compute gradients. The forward pass is the `(x ** 2).sum()`; in the backward pass it will do `2 * x`.

So `vmap(jacrev(f))(a)` is computing `(x ** 2).sum()`, then computing the jacobian (`2 * x`), and finally throwing away `(x ** 2).sum()` because you don't need it. This is a situation that torch.compile can help out with.

We don't support composing torch.compile and functorch transforms (vmap, jacrev) yet, but I did a quick experiment by tracing out what `vmap(jacrev(f))(a)` actually does under the hood and then applying torch.compile to that. torch.compile is able to speed up the expression to something close to just `2 * x` by pruning unused operations:
https://gist.github.com/zou3519/01fc60195b3dc55904e7ab22162e09bf

cc @kshitij12345 who is working on torch.compile support for functorch.

**Comment by FrankTianTT** [comment]
> We don't support composing torch.compile and functorch transforms (vmap, jacrev) yet, but I did a quick experiment by tracing out what `vmap(jacrev(f))(a)` actually does under the hood and then applying torch.compile to that. torch.compile is able to speed up the expression to something close to just `2 * x` by pruning unused operations:
> https://gist.github.com/zou3519/01fc60195b3dc55904e7ab22162e09bf

@zou3519 Thank you very much for your response!

The compiled `jacrev` is nearly as fast as, or even surpasses, the oracle function! However, I still found that the combined time of `df(a)` and `f(a)` is much faster than `vmap(jacrev(f))(a)`. This may be due to `jacrev(f)` containing too many unnecessary operations?

I would like to ask if there are any plans to support `jacrev` in `torch.compile`? A series of tasks require the first or second-order gradients of neural networks as output, rather than the output of the network itself. This is particularly relevant to physics-informed networks such as DeLaN and HNN. Enabling torch.compile to support jacrev would at least provide a tenfold improvement in speed, which is crucial for real-time prediction in physical environments.

Should I open another issue to formally state my request?

**Comment by zou3519** [comment]
Let's keep this issue open to track it
